--- wu-ftpd-2.6.2.orig/doc/examples/ftpconversions
+++ wu-ftpd-2.6.2/doc/examples/ftpconversions
@@ -1,9 +1,9 @@
  :.Z:  :  :/bin/compress -d -c %s:T_REG|T_ASCII:O_UNCOMPRESS:UNCOMPRESS
  :   : :.Z:/bin/compress -c %s:T_REG:O_COMPRESS:COMPRESS
- :.gz: :  :/bin/gzip -cd %s:T_REG|T_ASCII:O_UNCOMPRESS:GUNZIP
- :   : :.gz:/bin/gzip -9 -c %s:T_REG:O_COMPRESS:GZIP
- :   : :.tar:/bin/tar -c -f - %s:T_REG|T_DIR:O_TAR:TAR
- :   : :.tar.Z:/bin/tar -c -Z -f - %s:T_REG|T_DIR:O_COMPRESS|O_TAR:TAR+COMPRESS
- :   : :.tar.gz:/bin/tar -c -z -f - %s:T_REG|T_DIR:O_COMPRESS|O_TAR:TAR+GZIP
- :   : :.crc:/bin/cksum %s:T_REG::CKSUM
- :   : :.md5:/bin/md5sum %s:T_REG::MD5SUM
+ :.gz: :  :/bin/gzip -cd -- %s:T_REG|T_ASCII:O_UNCOMPRESS:GUNZIP
+ :   : :.gz:/bin/gzip -9 -c -- %s:T_REG:O_COMPRESS:GZIP
+ :   : :.tar:/bin/tar -c -f - -- %s:T_REG|T_DIR:O_TAR:TAR
+ :   : :.tar.Z:/bin/tar -c -Z -f - -- %s:T_REG|T_DIR:O_COMPRESS|O_TAR:TAR+COMPRESS
+ :   : :.tar.gz:/bin/tar -c -z -f - -- %s:T_REG|T_DIR:O_COMPRESS|O_TAR:TAR+GZIP
+ :   : :.crc:/bin/cksum -- %s:T_REG::CKSUM
+ :   : :.md5:/bin/md5sum -- %s:T_REG::MD5SUM
--- wu-ftpd-2.6.2.orig/src/access.c
+++ wu-ftpd-2.6.2/src/access.c
@@ -642,7 +642,7 @@
     }
 #ifndef NO_PRIVATE
     entry = (struct aclmember *) NULL;
-    if (getaclentry("private", &entry) && !strcasecmp(ARG0, "yes"))
+    if (getaclentry("private", &entry) && ARG0 && !strcasecmp(ARG0, "yes"))
 	priv_setup(_path_private);
 #endif /* !NO_PRIVATE */
 
@@ -650,7 +650,7 @@
     set = 0;
     while (!set && getaclentry("compress", &entry)) {
 	l_compress = 0;
-	if (!strcasecmp(ARG0, "yes"))
+	if (ARG0 && !strcasecmp(ARG0, "yes"))
 	    l_compress = 1;
 	for (which = 1; (which < MAXARGS) && ARG[which]; which++) {
 	    if (!wu_fnmatch(ARG[which], class, FNM_CASEFOLD)) {
@@ -664,7 +664,7 @@
     set = 0;
     while (!set && getaclentry("tar", &entry)) {
 	l_tar = 0;
-	if (!strcasecmp(ARG0, "yes"))
+	if (ARG0 && !strcasecmp(ARG0, "yes"))
 	    l_tar = 1;
 	for (which = 1; (which < MAXARGS) && ARG[which]; which++) {
 	    if (!wu_fnmatch(ARG[which], class, FNM_CASEFOLD)) {
@@ -678,23 +678,23 @@
 
     entry = (struct aclmember *) NULL;
     while (getaclentry("log", &entry)) {
-	if (!strcasecmp(ARG0, "commands")) {
-	    if (anonymous && strcasestr(ARG1, "anonymous"))
+	if (ARG0 && !strcasecmp(ARG0, "commands")) {
+	    if (ARG1 && anonymous && strcasestr(ARG1, "anonymous"))
 		log_commands = 1;
-	    if (guest && strcasestr(ARG1, "guest"))
+	    if (ARG1 && guest && strcasestr(ARG1, "guest"))
 		log_commands = 1;
-	    if (!guest && !anonymous && strcasestr(ARG1, "real"))
+	    if (ARG1 && !guest && !anonymous && strcasestr(ARG1, "real"))
 		log_commands = 1;
 	}
-	if (!strcasecmp(ARG0, "transfers")) {
+	if (ARG0 && !strcasecmp(ARG0, "transfers")) {
 	    set = 0;
-	    if (strcasestr(ARG1, "anonymous") && anonymous)
+	    if (ARG1 && strcasestr(ARG1, "anonymous") && anonymous)
 		set = 1;
-	    if (strcasestr(ARG1, "guest") && guest)
+	    if (ARG1 && strcasestr(ARG1, "guest") && guest)
 		set = 1;
-	    if (strcasestr(ARG1, "real") && !guest && !anonymous)
+	    if (ARG1 && strcasestr(ARG1, "real") && !guest && !anonymous)
 		set = 1;
-	    if (strcasestr(ARG2, "inbound"))
+	    if (ARG2 && strcasestr(ARG2, "inbound"))
 		inbound = 1;
 	    if (strcasestr(ARG2, "outbound"))
 		outbound = 1;
@@ -703,20 +703,20 @@
 	    if (set)
 		log_outbound_xfers = outbound;
 	}
-	if (!strcasecmp(ARG0, "security")) {
-	    if (strcasestr(ARG1, "anonymous") && anonymous)
+	if (ARG0 && !strcasecmp(ARG0, "security")) {
+	    if (ARG1 && strcasestr(ARG1, "anonymous") && anonymous)
 		log_security = 1;
-	    if (strcasestr(ARG1, "guest") && guest)
+	    if (ARG1 && strcasestr(ARG1, "guest") && guest)
 		log_security = 1;
-	    if (strcasestr(ARG1, "real") && !guest && !anonymous)
+	    if (ARG1 && strcasestr(ARG1, "real") && !guest && !anonymous)
 		log_security = 1;
 	}
-	if (!strcasecmp(ARG0, "syslog"))
+	if (ARG0 && !strcasecmp(ARG0, "syslog"))
 	    syslogmsg = 1;
-	if (!strcasecmp(ARG0, "xferlog"))
+	if (ARG0 && !strcasecmp(ARG0, "xferlog"))
 	    syslogmsg = 0;
-	if (!strcasecmp(ARG0, "syslog+xferlog")
-	    || !strcasecmp(ARG0, "xferlog+syslog"))
+	if (ARG0 && (!strcasecmp(ARG0, "syslog+xferlog")
+	    || !strcasecmp(ARG0, "xferlog+syslog")))
 	    syslogmsg = 2;
     }
 }
--- wu-ftpd-2.6.2.orig/src/ftpcount.c
+++ wu-ftpd-2.6.2/src/ftpcount.c
@@ -238,8 +238,8 @@
 #elif defined(sun)
 			sprintf(line, "/usr/ucb/ps auxww %ld", buf[which]);
 #else
-#if defined (LINUX_BUT_NOT_REDHAT_6_0)
-			sprintf(line, "/bin/ps axwww %d", buf[which]);
+#if defined (LINUX)
+			sprintf(line, "/bin/ps fwwwp %d", buf[which]);
 #else
 			sprintf(line, "/bin/ps -f -p %d", buf[which]);
 #endif
