diff -urN wu-ftpd-2.6.1.org/doc/examples/ftpaccess.heavy wu-ftpd-2.6.1/doc/examples/ftpaccess.heavy
--- wu-ftpd-2.6.1.org/doc/examples/ftpaccess.heavy	Sun Sep 10 15:26:26 2000
+++ wu-ftpd-2.6.1/doc/examples/ftpaccess.heavy	Sun Sep 10 18:34:41 2000
@@ -5,9 +5,9 @@
 class   local   real,guest,anonymous *.domain 0.0.0.0
 class   remote  real,guest,anonymous *
 
-limit   local   20  Any                 /etc/msgs/msg.toomany
-limit   remote  100 SaSu|Any1800-0600   /etc/msgs/msg.toomany
-limit   remote  60  Any                 /etc/msgs/msg.toomany
+limit   local   20  Any                 /etc/msgs/toomany
+limit   remote  100 SaSu|Any1800-0600   /etc/msgs/toomany
+limit   remote  60  Any                 /etc/msgs/toomany
 
 readme  README*    login
 readme  README*    cwd=*
@@ -26,7 +26,7 @@
 
 log commands real
 log transfers anonymous,real inbound,outbound
-shutdown /etc/shutmsg
+shutdown /etc/msgs/shutdown
 
 # all the following default to "yes" for everybody
 delete          no      guest,anonymous         # delete permission?
@@ -39,10 +39,10 @@
 upload  /var/ftp  *             no     nobody   nogroup 0000 nodirs
 upload  /var/ftp  /bin          no
 upload  /var/ftp  /etc          no
-upload  /var/ftp  /incoming     yes     root    daemon  0600 dirs
+upload  /var/ftp  /pub/Incoming     yes     root    daemon  0600 dirs
 
 # directory aliases...  [note, the ":" is not required]
-alias   inc:    /incoming
+alias   inc:    /pub/Incoming
 
 # cdpath
 cdpath  /incoming
@@ -50,8 +50,8 @@
 cdpath  /
 
 # path-filter...
-path-filter  anonymous  /etc/pathmsg  ^[-A-Za-z0-9_\.]*$  ^\.  ^-
-path-filter  guest      /etc/pathmsg  ^[-A-Za-z0-9_\.]*$  ^\.  ^-
+path-filter  anonymous  /etc/msgs/path  ^[-A-Za-z0-9_\.]*$  ^\.  ^-
+path-filter  guest      /etc/msgs/path  ^[-A-Za-z0-9_\.]*$  ^\.  ^-
 
 # specify which group of users will be treated as "guests".
 guestgroup ftponly
diff -urN wu-ftpd-2.6.1.org/doc/examples/ftphosts wu-ftpd-2.6.1/doc/examples/ftphosts
--- wu-ftpd-2.6.1.org/doc/examples/ftphosts	Sun Sep 10 15:26:26 2000
+++ wu-ftpd-2.6.1/doc/examples/ftphosts	Sun Sep 10 18:34:15 2000
@@ -3,5 +3,5 @@
 # Everything after a '#' is treated as comment,
 # empty lines are ignored
 
-    allow   bartm   somehost.domain
-    deny    fred    otherhost.domain 131.211.32.*
+#    allow   bartm   somehost.domain
+#    deny    fred    otherhost.domain 131.211.32.*
