--- wu-ftpd-2.4.2-vr17/src/extensions.c.pathname	Mon Apr 19 09:19:14 1999
+++ wu-ftpd-2.4.2-vr17/src/extensions.c	Mon Apr 19 09:19:37 1999
@@ -295,6 +295,7 @@
 #ifdef HAVE_GETCWD
                 (void) getcwd(outptr, MAXPATHLEN);
 #else
+#error	wu-ftpd on this platform has security deficiencies!!!
                 (void) getwd(outptr);
 #endif
                 break;
@@ -1202,6 +1203,14 @@
   *valid = 0;
   /* what's our current directory? */
 
+  /* XXX We could use dynamic RAM to store this path, but I'd rather just bail
+     out with an error. The rest of wu is so crufy that a long path might
+     just blow up later */
+
+  if ((strlen(name) + 1) > sizeof(path)) {
+    perror_reply (550, "Path too long");
+  }
+
   strcpy(path, name);
   sp = strrchr(path, '/');
   if (sp)
@@ -1383,6 +1392,14 @@
 
   /* what's our current directory? */
 
+  /* XXX We could use dynamic RAM to store this path, but I'd rather just bail
+     out with an error. The rest of wu is so crufy that a long path might
+     just blow up later */
+
+  if ((strlen(name) + 1) > sizeof(path)) {
+    perror_reply (550, "Path too long");
+  }
+
   strcpy (path, name);
   sp = strrchr (path, '/');
   if (sp)
@@ -2030,7 +2047,8 @@
     int i;
     int matches = 0;
 
-    strcpy(buf, patterns);
+    strncpy(buf, patterns, sizeof(buf));
+    buf[sizeof(buf)] = '\0';
     i = strlen(buf);
     buf[i++] = ',';
     buf[i++] = '\0';
@@ -2060,7 +2078,8 @@
     int i;
     int matches = 0;
 
-    strcpy(buf, patterns);
+    strncpy(buf, patterns, sizeof(buf));
+    buf[sizeof(buf)] = '\0';
     i = strlen(buf);
     buf[i++] = ',';
     buf[i++] = '\0';
@@ -2103,6 +2122,14 @@
     *bps = -1;
     *bpsmult = 1.0;
 
+  /* XXX We could use dynamic RAM to store this path, but I'd rather just bail
+     out with an error. The rest of wu is so crufy that a long path might
+     just blow up later */
+
+    if ((strlen(name) + 1) > sizeof(path)) {
+      perror_reply (550, "Path too long");
+    }
+
     /* what's our current directory? */
     strcpy(path, name);
     if (sp = strrchr(path, '/'))
@@ -2171,6 +2198,14 @@
     extern char *home;
     char *sp;
     int i;
+
+  /* XXX We could use dynamic RAM to store this path, but I'd rather just bail
+     out with an error. The rest of wu is so crufy that a long path might
+     just blow up later */
+
+    if ((strlen(name) + 1) > sizeof(path)) {
+      perror_reply (550, "Path too long");
+    }
 
     /* what's our current directory? */
     strcpy(path, name);
